File: src/game/step.ts
import {
  type Action,
  type Edge,
  type Flow,
  type GameState,
  type Owner,
  MAX_STRENGTH,
  MIN_STRENGTH_TO_SEND,
  REGEN_PER_SEC,
  TRANSFER_PER_SEC,
} from './state';

type Force = number[]; // strength contribution per player at one node

const zeroForce = (n: number): Force => Array(n).fill(0);

const adjacencyCache = new WeakMap<readonly Edge[], Set<number>>();

function edgeKey(a: number, b: number): number {
  return a < b ? a * 1e7 + b : b * 1e7 + a;
}

function adjacencySet(edges: readonly Edge[]): Set<number> {
  let s = adjacencyCache.get(edges);
  if (s) return s;
  s = new Set();
  for (const e of edges) s.add(edgeKey(e.a, e.b));
  adjacencyCache.set(edges, s);
  return s;
}

export function applyAction(state: GameState, action: Action): GameState {
  if (action.kind !== 'toggleFlow') return state;
  const src = state.nodes[action.src];
  if (!src || src.owner !== action.player) return state;
  if (!adjacencySet(state.edges).has(edgeKey(action.src, action.dst))) return state;
  const onEdge = state.flows.findIndex(f =>
    (f.src === action.src && f.dst === action.dst) ||
    (f.src === action.dst && f.dst === action.src),
  );
  if (onEdge < 0) {
    return { ...state, flows: [...state.flows, { src: action.src, dst: action.dst, player: action.player }] };
  }
  const existing = state.flows[onEdge];
  const exact = existing.src === action.src && existing.dst === action.dst && existing.player === action.player;
  const flows = state.flows.filter((_, j) => j !== onEdge);
  if (exact) return { ...state, flows };
  return { ...state, flows: [...flows, { src: action.src, dst: action.dst, player: action.player }] };
}

export function step(state: GameState, dt: number): GameState {
  const N = state.numPlayers;
  const forces: Force[] = state.nodes.map(() => zeroForce(N));

  for (const node of state.nodes) {
    if (node.owner !== null) forces[node.id][node.owner] += REGEN_PER_SEC * dt;
  }

  const aliveFlows: Flow[] = [];
  for (const flow of state.flows) {
    const src = state.nodes[flow.src];
    if (src.owner !== flow.player) continue;
    aliveFlows.push(flow);
    if (src.strength < MIN_STRENGTH_TO_SEND) continue;
    const k = TRANSFER_PER_SEC * dt;
    forces[flow.src][flow.player] -= k;
    forces[flow.dst][flow.player] += k;
  }

  const nodes = state.nodes.map(node => {
    const force = forces[node.id];
    let owner = node.owner;
    let strength = node.strength;

    if (owner !== null) {
      strength += force[owner];
      for (let p = 0; p < N; p++) if (p !== owner) strength -= force[p];
    } else {
      for (let p = 0; p < N; p++) if (force[p] > 0) strength -= force[p];
    }

    if (strength < 0) {
      let bestP: Owner = null, best = 0;
      for (let p = 0; p < N; p++) {
        if (p !== owner && force[p] > best) { best = force[p]; bestP = p; }
      }
      if (bestP !== null) { owner = bestP; strength = -strength; }
      else strength = 0;
    }

    if (strength > MAX_STRENGTH) strength = MAX_STRENGTH;
    if (strength < 0) strength = 0;

    return { ...node, owner, strength };
  });

  return { ...state, nodes, flows: aliveFlows, tick: state.tick + 1 };
}

File: python/flux/step.py
from __future__ import annotations
from typing import List, Optional
from .state import (
    Action,
    Edge,
    Flow,
    GameNode,
    GameState,
    ATTACK_BONUS,
    MAX_STRENGTH,
    MIN_STRENGTH_TO_SEND,
    REGEN_PER_SEC,
    TRANSFER_PER_SEC,
)

# Per-edges-tuple adjacency cache, keyed by id() (object identity), like the
# JS WeakMap keyed on the edges array.
_adjacency_cache: dict[int, set[int]] = {}


def _edge_key(a: int, b: int) -> int:
    return a * 10_000_000 + b if a < b else b * 10_000_000 + a


def _adjacency_set(edges: tuple[Edge, ...]) -> set[int]:
    k = id(edges)
    s = _adjacency_cache.get(k)
    if s is not None:
        return s
    s = set()
    for e in edges:
        s.add(_edge_key(e.a, e.b))
    _adjacency_cache[k] = s
    return s


def apply_action(state: GameState, action: Action) -> GameState:
    if action.kind != "toggleFlow":
        return state
    if action.src < 0 or action.src >= len(state.nodes):
        return state
    src = state.nodes[action.src]
    if src.owner != action.player:
        return state
    if _edge_key(action.src, action.dst) not in _adjacency_set(state.edges):
        return state
    on_edge = -1
    for i, f in enumerate(state.flows):
        if (f.src == action.src and f.dst == action.dst) or (
            f.src == action.dst and f.dst == action.src
        ):
            on_edge = i
            break
    if on_edge < 0:
        return GameState(
            nodes=state.nodes,
            edges=state.edges,
            flows=state.flows + (Flow(src=action.src, dst=action.dst, player=action.player),),
            tick=state.tick,
            num_players=state.num_players,
        )
    existing = state.flows[on_edge]
    exact = (
        existing.src == action.src
        and existing.dst == action.dst
        and existing.player == action.player
    )
    flows = tuple(f for j, f in enumerate(state.flows) if j != on_edge)
    if exact:
        return GameState(
            nodes=state.nodes,
            edges=state.edges,
            flows=flows,
            tick=state.tick,
            num_players=state.num_players,
        )
    return GameState(
        nodes=state.nodes,
        edges=state.edges,
        flows=flows + (Flow(src=action.src, dst=action.dst, player=action.player),),
        tick=state.tick,
        num_players=state.num_players,
    )


def step(state: GameState, dt: float) -> GameState:
    N = state.num_players
    n_nodes = len(state.nodes)
    forces: List[List[float]] = [[0.0] * N for _ in range(n_nodes)]

    for node in state.nodes:
        if node.owner is not None:
            forces[node.id][node.owner] += REGEN_PER_SEC * dt

    alive_flows: List[Flow] = []
    for flow in state.flows:
        src = state.nodes[flow.src]
        if src.owner != flow.player:
            continue
        alive_flows.append(flow)
        if src.strength < MIN_STRENGTH_TO_SEND:
            continue
        k = TRANSFER_PER_SEC * dt
        forces[flow.src][flow.player] -= k
        enemy = state.nodes[flow.dst].owner != flow.player
        forces[flow.dst][flow.player] += k * (1.0 + ATTACK_BONUS) if enemy else k

    new_nodes: List[GameNode] = []
    for node in state.nodes:
        force = forces[node.id]
        owner: Optional[int] = node.owner
        strength = node.strength

        if owner is not None:
            strength += force[owner]
            for p in range(N):
                if p != owner:
                    strength -= force[p]
        else:
            for p in range(N):
                if force[p] > 0:
                    strength -= force[p]

        if strength < 0:
            best_p: Optional[int] = None
            best = 0.0
            for p in range(N):
                if p != owner and force[p] > best:
                    best = force[p]
                    best_p = p
            if best_p is not None:
                owner = best_p
                strength = -strength
            else:
                strength = 0.0

        if strength > MAX_STRENGTH:
            strength = MAX_STRENGTH
        if strength < 0:
            strength = 0.0

        new_nodes.append(GameNode(id=node.id, pos=node.pos, owner=owner, strength=strength))

    return GameState(
        nodes=tuple(new_nodes),
        edges=state.edges,
        flows=tuple(alive_flows),
        tick=state.tick + 1,
        num_players=state.num_players,
    )
